All: Document classes option

jzaefferer opened the pull request:
This document the default `classes` option and the `_addClass` method on `$.Widget`, and the `classes` option on accordion and draggable. There's a bunch of issues that need to be addressed. I'm including screenshots here to make previewing easier.

### `classes` option on $.Widget:

![...](http://bassistance.de/i/fc4b6f.png)

Okay, but not very useful. Looking for suggestions how to improve this.

### `_addClass` method:

![...](http://bassistance.de/i/4494a6.png)

* Apparently `<method>` only supports a single `<example>` element, so the three descriptions are all mashed together, same for the code snippets. I think having multiple examples here is important enough to justify a change to grunt-jquery-content.
* Multiple descriptions are rather bad, like "The element to add the classes to". Looking for suggestions for improving these.
* I could make the reference to the `classes` option a link, but since that references this method, we end up with circular references without adding any value. Ignore?

### Accordion `classes` option:

![...](http://bassistance.de/i/ef918b.png)

* The `Default` has to be specified as an attribute, so it can't be formatted properly, so it looks really bad. Only alternative I can think of is to put the default into the `<desc>` element.
* Is the description okay? If not, suggestions for improving it?
* I've used a value of `null` for the example. I'm not sure if that is a good idea. See below for what I've used for draggable.

### Draggable `classes` option:

![...](http://bassistance.de/i/b806f2.png)

* The default value isn't quite as bad as for accordion, but still wraps over multiple lines.
* Same description as for accordion - can/should we somehow share that?
* In the example I'm pretending to add an extra class to the "draggable-handle". Is that good enough?

---

Once these issues are addressed, I can add the docs for `_removeClass`, `_toggleClass` and the remaining widgets.

arschmitz commented:
A couple general things i noticed right off. In the widget documentation, it never mentions the concept of the option, which is to map structural classes to theme classes, that might help understanding. Second in the widget documentation the auto destroy functionality is never mentioned we want to include that for sure. 

One thing that might help clear up the difference between keys, values, and extra along with their proper usage is to explain how we use them and the naming convention we use. 

* Keys - All classes for an individual widget are prefixed with the widget name ex: "ui-accordion" so any class starting with `ui-accordion*` is a key, its structural and specific to that widget these classes are always added no matter what. 
* Values - Theme classes that do things like add colors adjust border radius etc these classes are purely presentational and do not effect the function of the widget if missing. These classes are configurable
* Extra - these classes are essential to the widget operation and or layout but are not specific to the widget. These classes are added no matter what.

On the example of getting and setting, we want to make sure that people understand if they pass an object, it will over ride the whole option,  not just the values they pass. If we add an example of a deep option getter/ setter this should help clear that up. 

> I could make the reference to the classes option a link, but since that references this method, we end up with circular references without adding any value. Ignore?"

I think this still adds value because it avoids scrolling back and forth when your trying to understand the whole flow of the added classes stuff

> The Default has to be specified as an attribute, so it can't be formatted properly, so it looks really bad. Only alternative I can think of is to put the default into the <desc> element.

Can we parse this and try to do some type checking? Kind of like what core does when you use `.data` on an attribute and it tries to figure out the type and return the correct type of data? @scottgonzalez is this at all possible to do with the xslt? if not i would change it to just say object and put it in the description I think. 

> Same description as for accordion - can/should we somehow share that?

Yes i think we should write a single description and share on every widget it will never change

jzaefferer commented:
Updated, with lots of improvements. Take a look, I think I've addressed everything you commented on. Smashed examples are still a problem.

jzaefferer commented:
Once I got a review of my last update, https://github.com/jquery/grunt-jquery-content/pull/63 is merged and published, and [the full list of defaults is reviewed](https://github.com/jquery/api.jqueryui.com/issues/256#issuecomment-83991811), I'll add the remaining methods and update the other widgets.

arschmitz commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+					</ul>
+					<p>In addition to the structural and theme classes, there are also extra classes that are required for layout or other reasons and can't be removed. These are specified in the <code>extras</code> argument to <code>_addClass()</code>, but aren't part of this option.</p>
+					<p>As long as the <code>_addClass()</code>, <code>_removeClass()</code> and <code>_toggleClass()</code> methods are used, any added classes are automatically removed when the widget is destroyed. This can heavily simplify the implementation of custom <code>_destroy()</code> methods.</p>
+					<p class="warning">Setting the <code>classes</code> option will override all default properties. To only change specific values, use deep setters, e.g. <code>.option( "classes.ui-accordion-header", null )</code>.</p>
```
This is only AFTER init when using the getter / setter methods, on init its a deep extend. Might want to make that clear here.

arschmitz commented on includes/classes-option-desc.xml:
```diff
@@ -0,0 +1,5 @@
+<desc>
+	<p>Specify additional classes to add to the widget's elements. Any of the default keys can be specified to override their value. To learn more about this option, check out the documentation for <a href="/jQuery.widget/#option-classes">jQuery.Widget's <code>classes</code> option</a>.</p>
+	<p>The default value for the <placeholder name="name"/> widget:</p>
+</desc>
```
Missing line at EOF

jzaefferer commented:
@arschmitz I've addressed both your comments and fixed the grunt-jquery-content issue. If you could take another look and confirm my list ( https://github.com/jquery/api.jqueryui.com/issues/256#issuecomment-83991811 ), I can write the rest.

arschmitz commented:
@jzaefferer have not had a chance to go through all of this yet but something i wanted to put down so i don't forget

1.) we have added new `ui-widgetname-*` classes to some widgets so we want to make sure those get added to the theme sections where appropriate. For example selectmenu now has `selectmenu-button-open` and `selectmenu-button-closed` ( though it appears selectmenu has no theme section at all )

2.) While working on the markup structure tests @scottgonzalez and i had talked about if maybe we should also add the `ui-widget-*` classes to this section thoughts?.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
 			<xi:include href="../includes/widget-option-show.xml" xmlns:xi="http://www.w3.org/2003/XInclude"/>
+			<option name="classes" default="{}" example-value='{ "custom-header": "icon-warning" }'>
+				<desc>
+					<p>Extra classes to add to the widget. See the <a href="#method-_addClass"><code>_addClass()</code> method</a> for using this in custom widegts. Check out the documentation of individual widgets to see which classes they support.</p>
```
Typo: widegts

scottgonzalez replied:
Perhaps provide more context upfront:

Additional (thematic) classes to add to the widget, in addition to the structural classes.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+			<option name="classes" default="{}" example-value='{ "custom-header": "icon-warning" }'>
+				<desc>
+					<p>Extra classes to add to the widget. See the <a href="#method-_addClass"><code>_addClass()</code> method</a> for using this in custom widegts. Check out the documentation of individual widgets to see which classes they support.</p>
+					<p>The primary motivation of this option is to map structural classes to theme classes. In other words, any class prefixed with namespace and widget, like <code>"ui-accordion-"</code>, is considered a structural class. These are always added to the widget. In contrast to that, any class not specific to the widget is considered a theme class. These could be part of jQuery UI's CSS framework,  ...
```
I'm pretty sure we've consistent in using progressbar for all of our generic examples.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+					<ul>
+						<li>structurual classes: the keys of this option, equivalent to the <code>keys</code> argument to <code>_addClass()</code>, for example <code>"ui-accordion-header"</code></li>
+						<li>theme classes: the values of this option, for example <code>"ui-corner-top"</code></li>
+					</ul>
```
This seems a bit out of place in this format. Perhaps just add to the introductory paragraph instead. Something like "The structural classes are used as keys to the `classes` option and the thematic classes are the values."

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+						<li>structurual classes: the keys of this option, equivalent to the <code>keys</code> argument to <code>_addClass()</code>, for example <code>"ui-accordion-header"</code></li>
+						<li>theme classes: the values of this option, for example <code>"ui-corner-top"</code></li>
+					</ul>
+					<p>In addition to the structural and theme classes, there are also extra classes that are required for layout or other reasons and can't be removed. These are specified in the <code>extras</code> argument to <code>_addClass()</code>, but aren't part of this option.</p>
```
This doesn't belong in the `classes` option documentation.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+						<li>theme classes: the values of this option, for example <code>"ui-corner-top"</code></li>
+					</ul>
+					<p>In addition to the structural and theme classes, there are also extra classes that are required for layout or other reasons and can't be removed. These are specified in the <code>extras</code> argument to <code>_addClass()</code>, but aren't part of this option.</p>
+					<p>As long as the <code>_addClass()</code>, <code>_removeClass()</code> and <code>_toggleClass()</code> methods are used, any added classes are automatically removed when the widget is destroyed. This can heavily simplify the implementation of custom <code>_destroy()</code> methods.</p>
```
I don't think this belongs here either.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -192,6 +192,21 @@
+					</ul>
+					<p>In addition to the structural and theme classes, there are also extra classes that are required for layout or other reasons and can't be removed. These are specified in the <code>extras</code> argument to <code>_addClass()</code>, but aren't part of this option.</p>
+					<p>As long as the <code>_addClass()</code>, <code>_removeClass()</code> and <code>_toggleClass()</code> methods are used, any added classes are automatically removed when the widget is destroyed. This can heavily simplify the implementation of custom <code>_destroy()</code> methods.</p>
+					<p class="warning">Setting the <code>classes</code> option after creation will override all default properties. To only change specific values, use deep setters, e.g. <code>.option( "classes.ui-accordion-header", null )</code>.</p>
```
Not just default values, but any existing values, even if they've been customized. Perhaps we should just automatically handle this for options with a type of Object.

arschmitz replied:
That might be a good idea i think this is going to lead to a lot of bug reports and confusion when someone sets a single value and wipes out the whole object. We have not had a lot of big object options like this in the past for it to really be an issue

jzaefferer replied:
Handling this for all `Object` type options sounds good to me. Let's deal with that together with the display of default values.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +215,43 @@
+				</argument>
+				<argument name="keys" type="String">
+					<desc>
+						The classes to add. If a property of the <code>classes</code> option matches a key, the value will be added as well.
```
Should probably clarify that it's as space-delimited list of classes.

scottgonzalez replied:
Link to classes option.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +215,43 @@
+				<argument name="keys" type="String">
+					<desc>
+						The classes to add. If a property of the <code>classes</code> option matches a key, the value will be added as well.
+						<p>Like the <code>element</code> argument, this argument is also optional. To skip it, specify <code>null</code>.</p>
```
Not actually optional. We should probably add a note about why you would use the method with a null key.

scottgonzalez replied:
Perhaps the note isn't necessary since we explicitly mention this in the description for `extra`.

jzaefferer replied:
I'll update the text to just talk about using `null` for skipping and adding an explicit reference to the other argument.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +215,43 @@
+					<desc>Add the "loading-indicator" class to the widget's element (<code>this.element</code>. Will also add any additional classes specified through the <code>classes</code> option for the given class.</desc>
+					<code><![CDATA[
+this._addClass( "loading-indicator" );
+]]></code>
```
This contradicts the earlier sentence 'In other words, any class prefixed with namespace and widget, like "ui-accordion-"...' and it isn't a very realistic case for widgets that follow our pattern. Perhaps pick a class from progressbar.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +215,43 @@
+]]></code>
+				</example>
+				<example>
+					<desc>Add the "popup-header" class to the specified element (here referencing <code>this.popup</code>). Will also add any additional classes specified through the <code>classes</code> option for the given class. In addition, it will always add the "ui-front" class.</desc>
```
Add a namespace to the class, I think we use a "demo" namespace elsewhere.

scottgonzalez commented on includes/classes-option-desc.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<desc>
+	<p>Specify additional classes to add to the widget's elements. Any of the default keys can be specified to override their value. To learn more about this option, check out the documentation for <a href="/jQuery.widget/#option-classes">jQuery.Widget's <code>classes</code> option</a>.</p>
```
Probably better to point to a detailed learn article than the API docs. See https://github.com/jquery/learn.jquery.com/pull/574.

scottgonzalez commented on entries/accordion.xml:
```diff
@@ -86,6 +86,21 @@
 				</desc>
 			</type>
 		</option>
+		<option name="classes" type="Object" default='Object' example-value='{ "ui-accordion-header": null }'>
```
This isn't a great example value because it looks like it's just removing the value from one key, but it's effectively removing the value from all keys.

jzaefferer replied:
Well, for the getter/setter example, not for the init example, right? Even with something other then `null` we'd still overwrite the other properties when using this, right?

```js
.accordion( "option", "classes", { "ui-accordion-header": "whatever" } );
```

arschmitz replied:
@jzaefferer that would but you can do 
```js
.accordion( "option", "classes.ui-accordion-header", "whatever" );
```

scottgonzalez replied:
Yeah, but we can just create custom examples that use `"classes.ui-accordion-header"` instead of `"classes"`.

jzaefferer replied:
How's this?

![...](http://bassistance.de/i/a10a00.png)

The styling for `<example>` element doesn't match the one for the generated examples. I guess one more TODO for grunt-jquery-content to add to the list.

arschmitz replied:
I would change `classes` in the description of the 2nd and 3rd examples to specify that your specifically setting the value for the `ui-accordion-header` property on the classes option

jzaefferer commented:
Updated, addressing Scott's review and the last comment from Alex. Preview:

![...](http://bassistance.de/i/6a1b68.png)

This depends on https://github.com/jquery/grunt-jquery-content/pull/67 for the styling of `<example>` elements inside `<option>`. It also improves headers and description for the "Examples" section (same "Examples" header, but displays description as regular paragraph).

I'd like to avoid having to create custom examples for each widget, but that's currently blocked by https://github.com/jquery/grunt-jquery-content/issues/66

arthurvr commented on includes/classes-option-example.xml:
```diff
@@ -0,0 +1,23 @@
+$( ".selector" ).<placeholder name="name"/>( "option", "classes.ui-<placeholder name="name"/>", "highlight" );
+</code>
+	</example>
+</examples>
```
missing newline at EOF

jzaefferer commented:
Added documentation for `_removeClass()` and `_toggleClass()`. Improved docs for `_addClass()` as well. Updated examples for classes option, where both accordion and draggable use the same include. Landed all the necessary improvements in grunt-jquery-content 2.2.0, no blocker left.

So if this looks good and [my full list of details](https://github.com/jquery/api.jqueryui.com/issues/256#issuecomment-83991811) is verified, all that's left is adding the `<option>` with the right `<default>` and the two includes to the other widgets.

scottgonzalez commented:
I still think the default value should be the actual default value (i.e., none of the `undefined` values should exist). We should move the full structure up to the Theming section and reference it from the `classes` option description.

scottgonzalez commented:
Let's update the `classes` option examples for `jQuery.Widget` so it shows setting the full value and a single value.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +209,106 @@
+				<desc>
+					Add classes to an element of the widget.
+					<p>This provides a hook for the user to add additional classes or replace default styling classes, through the <a href="#option-classes"><code>classes</code> option</a>.</p>
+					<p>It also provides automatic removal of these classes when a widget is destroyed, as long as you're using <code>_addClass()</code>, <code>_removeClass()</code> and <code>_toggleClass()</code> together. This can heavily simplify the implementation of custom <code>_destroy()</code> methods.</p>
```
Let's add links to the methods.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +209,106 @@
+					</desc>
+				</argument>
+				<argument name="extra" type="String" optional="true">
+					<desc>Additional classes to add, required for layout or other reasons. Unlike the <code>keys</code> argument, these aren't associated with any properties of the <code>classes</code> option. Just like <code>keys</code>, they will also be automatically removed when destroying the widget.</desc>
```
Link `classes` option.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +209,106 @@
+					<desc>Additional classes to add, required for layout or other reasons. Unlike the <code>keys</code> argument, these aren't associated with any properties of the <code>classes</code> option. Just like <code>keys</code>, they will also be automatically removed when destroying the widget.</desc>
+				</argument>
+				<example>
+					<desc>Add the <code>ui-progressbar</code> class to the widget's element (<code>this.element</code>). Will also add any additional classes specified through the <code>classes</code> option for the given class.</desc>
```
Link `classes` option.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +209,106 @@
+]]></code>
+				</example>
+				<example>
+					<desc>Add the <code>demo-popup-header</code> class to the specified element (here referencing <code>this.popup</code>). Will also add any additional classes specified through the <code>classes</code> option for the given class. In addition, it will always add the <code>ui-front</code> class.</desc>
```
Link `classes` option.

scottgonzalez commented:
The missing links also apply to the other methods.

scottgonzalez commented on entries/jQuery.widget.xml:
```diff
@@ -200,6 +209,106 @@
+					<desc>Additional classes to toggle, required for layout or other reasons. Unlike the <code>keys</code> argument, these aren't associated with any properties of the <code>classes</code> option. Just like <code>keys</code>, they will also be automatically removed when destroying the widget.</desc>
+				</argument>
+				<argument name="add" type="Boolean">
+					<desc>Indicates whether to add or remove the specified classes, where a boolean <code>true</code> indicates that classes should be added, a boolean <code>false</code> indicates that classes should be removed.</desc>
```
Even though the docs show that this is required, it's probably worth an explicit mention that unlike `.toggleClass()`, this method requires the boolean.

Thoughts on whether this should just be a regular sentence in the description or highlighted as a warning?

arschmitz replied:
Personally i would say this is a big enough deviation from the potentially "expected" behavior that it merits a warning

jzaefferer replied:
I've added a warning to the method description.

jzaefferer commented:
Added three separate commits, that should address everything. For the `add` argument warning, I've added that to the `<desc>` of the `_toggleClass()` method.

scottgonzalez commented:
The updates look good. I'll do another full review to see how everything looks at this point.

scottgonzalez commented:
Did another round of review and everything looked good. I guess the next step is updating the individual widget docs for the remaining widgets?

arschmitz commented:
I just finished reviewing again as well and everything looks good to me.

jzaefferer commented:
Updated to include docs for all components. Turned out that theming information was missing or incomplete in many places. That was responsible for most of the effort.

jzaefferer commented:
Updated, addressed all of Scott's comments. I've changed the nesting for `ui-selectee` to go under `ui-selectable`, and both `ui-sortable-handle` and `ui-sortable-placeholder` under `ui-sortable`.

scottgonzalez commented:
Looks good.

arschmitz commented:
LGTM :+1:

jzaefferer commented:
Turns out I based this on master, instead of 1-12, and sent this PR against master as well. Will rebase against 1-12 and create a new PR with squashed commits.
